src/numerics/funcs.cpp

```diff
@@ -27,4 +29,88 @@ doublereal linearInterp(doublereal x, const vector_fp& xpts,
     return ff;
 }
 
+double trapezoidal(const Eigen::ArrayXd& f, const Eigen::ArrayXd& x)
+{
+    // check length
+    if (f.size() != x.size()) {
+        throw CanteraError("trapezoidal",
+                           "Vector lengths need to be the same.");
+    }
+    // Vector of f(i+1) + f(i)
+    Eigen::VectorXd f_av = f.tail(f.size() - 1) + f.head(f.size() - 1);
+    // Vector of x(i+1) - x(i)
+    Eigen::VectorXd x_diff = x.tail(x.size() - 1) - x.head(x.size() - 1);
+    // check if the coordinate is a monotonically increase vector.
+    if ((x_diff.array() <= 0.0).any()) {
+        throw CanteraError("trapezoidal",
+            "x (coordinate) needs to be the monotonically increasing.");
+    }
+    return f_av.dot(x_diff) / 2.0;
+}
+
+double basicSimpson(const Eigen::ArrayXd& f, const Eigen::ArrayXd& x)
+{
+    if (f.size() != x.size()) {
+        throw CanteraError("basicSimpson",
+                           "Vector lengths need to be the same.");
+    }
+    if (f.size() < 2) {
+        throw CanteraError("basicSimpson",
+                           "Vector lengths need to be larger than two.");
+    }
+    if (f.size()%2 == 0) {
+        throw CanteraError("basicSimpson",
+                           "Vector lengths need to be an odd number.");
+    }
+
+    size_t N = f.size() - 1;
+    Eigen::VectorXd h = x.tail(N) - x.head(N);
+
+    double sum = 0.0;
+    for (size_t i = 1; i < N; i+=2) {
+        double h0 = h[i-1];
+        double h1 = h[i];
+        double hph = h1 + h0;
+        double hdh = h1 / h0;
+        double hmh = h1 * h0;
+        sum += (hph / 6.0) * (
+                    (2.0 - hdh) * f[i - 1] + (pow(hph, 2) / hmh) * f[i] +
+                    (2.0 - 1.0 / hdh) * f[i + 1]);
+    }
+    return sum;
+}
+
+double simpson(const Eigen::ArrayXd& f, const Eigen::ArrayXd& x)
+{
+    if (f.size()%2 == 1) {
+        return basicSimpson(f, x);
+    } else {
+        size_t N = f.size() - 1;
+        // pick first N-1 point for simpson
+        double headSimps = basicSimpson(f.head(N), x.head(N));
+        // pick last N-1 point for simpson
+        double tailSimps = basicSimpson(f.tail(N), x.tail(N));
+        // Use trapezoidal rules for the first interval
+        double headTrap = 0.5 * (x[1] - x[0]) * (f[1] + f[0]);
+        // Use trapezoidal rules for the last interval
+        double tailTrap = 0.5 * (x[N] - x[N-1]) * (f[N] + f[N-1]);
+        return (headSimps + tailSimps + headTrap + tailTrap) / 2.0;
```

Comment:
Where does this method for dealing with an even number of points come from? It seems like kind of a surprising extra computational expense, having to calculate Simpson's rule twice for almost the whole interval.

Reply:
This is from scipy.simpson default, https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.simpson.html. However, I agree with you that this is not efficient.